### PR TITLE
[#10] [Integration] As a user, I can sign in with email and password 

### LINKED ISSUE #10: [Integrate] As a user, I can sign in with email and password
## Why
The user can log in with their credentials.

## Acceptance Criteria
- The API endpoint `/api/v1/oauth/token` is called.

  - If the API returns success, the user will navigate to the Home screen (which can be blank for now)
  - If the API returns an error, the user will see a custom error popup (which will be handled in a separate story)
- Unit tests are added and passed.
- The app must save the `access_token`, `token_type`, and the `refresh_token` on the local storage after successful login

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
- Closes #10 

## What happened 👀

- Integrate login
- Save tokens after successful login
- Add unit test

## Insight 📝

- Added `ErrorResponse` and `ErrorListResponse` for custom error display

## Proof Of Work 📹

https://user-images.githubusercontent.com/32578035/237045035-0b77abc2-f494-4442-9848-aa67f189a065.mov

https://user-images.githubusercontent.com/32578035/237045117-86f7536c-1054-4738-b9e4-2ea60ef62637.mov

### CODE REVIEW

**Comment by manh-t** [comment]
I face an error with Login Form. Steps:
- Press Login button at the first time.
- Then enter any email and password.
- Press Login button again.
- Observe that there is a red square box displayed.

Please help me check. I think this error is because the wrong usage `ref.read` or `ref.watch` in your `_emailValidator` and `_passwordValidator` methods.

https://github.com/nimblehq/flutter-ic-kaung-thieu/assets/60863885/abd03334-c95f-4291-b236-b4a64fb6a0f8

File: pubspec.yaml [hunk 1]
```diff
@@ -39,6 +39,7 @@ dependencies:
   flutter_secure_storage: ^8.0.0
   flutter_config_plus: ^1.1.1
   page_view_dot_indicator: ^2.1.0
+  collection: ^1.17.0
```
[the remark this answers is not in the document]
Comment (kaungkhantsoe) [reply pubspec.yaml:42]: @manh-t I tried using `firstOrNull` from `collection` before adding this and I couldn't. That's why. Any idea why I couldn't use `firstOrNull` before?

File: lib/screens/login/login_form.dart [hunk 2]
```diff
@@ -86,43 +109,55 @@ class _LoginFormState extends State<LoginForm> {
...
+          error: (error, _) {
+            showAlertDialog(
+                context: context,
+                title: _localizations.loginFailAlertTitle,
+                message:
+                    (error as String?) ?? _localizations.loginFailAlertMessage,
+                actions: [
+                  TextButton(
+                    child: Text(_localizations.okText),
+                    onPressed: () => Navigator.pop(context),
```
[the remark this answers is not in the document]
Comment (kaungkhantsoe) [reply lib/screens/login/login_form.dart:133]: `context.pop` doesn't work @manh-t . It doesn't close the pop-up. 😢

File: pubspec.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (manh-t) [reply pubspec.yaml:42]: @kaungkhantsoe maybe the IDE didn't suggest importing `collection` package, we have to import it manually

File: lib/screens/login/login_form.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (manh-t) [reply lib/screens/login/login_form.dart:133]: ah I see, this is the dialog. Let's keep it 👌

File: pubspec.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (manh-t) [reply pubspec.yaml:42]: @kaungkhantsoe it's not the error, it's the warning and we are fine with it. After removing `collection` from pubspec, please fetch the pubpsec file again.

File: lib/screens/login/login_form.dart [hunk 3]
```diff
@@ -59,70 +106,68 @@ class _LoginFormState extends State<LoginForm> {
...
+        _passwordTextField,
+        const SizedBox(
+          height: fieldSpacing,
+        ),
+        SizedBox(
+          height: 56,
+          width: double.infinity,
+          child: _loginButton,
+        )
+      ],
```
Comment (manh-t) [inline lib/screens/login/login_form.dart:170]: @kaungkhantsoe I realize ... we should follow the [best practices](https://docs.flutter.dev/cookbook/forms/validation) from Flutter to create a Form 😅 Currently you are creating multiple work around for the form validator. Why don't we follow Flutter instruction to create a Form as the simplest way? 🤤 
I have modified the code and it works well. Here is the example:
```suggestion
    return Form(
      key: _formKey,
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          TextFormField(
            keyboardType: TextInputType.emailAddress,
            autocorrect: false,
            decoration: FormFieldDecoration(
              hint: _localizations.emailInputHint,
              hintTextStyle: _textTheme.bodyLarge,
            ),
            style: _textTheme.bodyLarge,
            controller: emailController,
            validator: (value) => ref
                .read(loginViewModelProvider.notifier)
                .checkEmail(value, _localizations.invalidEmailError),
          ),
          const SizedBox(
            height: fieldSpacing,
          ),
          TextFormField(
            autocorrect: false,
            obscureText: true,
            decoration: FormFieldDecoration(
              hint: _localizations.passwordInputHint,
              hintTextStyle: _textTheme.bodyLarge,
            ),
            style: _textTheme.bodyLarge,
            controller: passwordController,
            validator: (value) => ref
                .read(loginViewModelProvider.notifier)
                .checkPassword(value, _localizations.invalidPasswordError),
          ),
          const SizedBox(
            height: fieldSpacing,
          ),
          SizedBox(
            height: 56,
            width: double.infinity,
            child: _loginButton,
          )
        ],
      ),
```

File: lib/screens/login/login_form.dart [hunk 4]
```diff
@@ -59,70 +106,68 @@ class _LoginFormState extends State<LoginForm> {
...
-              child: Text(_localizations.okText),
-              onPressed: () => Navigator.pop(context),
-            ),
-          ]);
+    final isReadyToLogin =
+        (ref.watch(isValidEmailStreamProvider).value ?? false) &&
+            (ref.watch(isValidPasswordStreamProvider).value ?? false);
+    KeyboardManager.dismiss(context);
+    if (isReadyToLogin) {
+      ref.read(loginViewModelProvider.notifier).login();
```
Comment (manh-t) [inline lib/screens/login/login_form.dart:128]: Then we don't need the email or password stream variable here. Just check it by using `_formKey`:
```suggestion
    KeyboardManager.dismiss(context);
    if (_formKey.currentState!.validate()) {
      ref.read(loginViewModelProvider.notifier).login();
```

File: lib/screens/login/login_form.dart [hunk 5]
```diff
@@ -1,48 +1,95 @@
...
+  }
+
+  String? _passwordValidator() {
+    final isValidPassword =
+        ref.watch(isValidPasswordStreamProvider).value ?? false;
+    if (!isValidPassword && _isFormSubmitted) {
+      return _localizations.invalidPasswordError;
+    }
+    return null;
+  }
```
Comment (manh-t) [inline lib/screens/login/login_form.dart:67]: we can get rid of unused stuff here by following the best practice's approach

File: lib/screens/login/login_form.dart [hunk 6]
```diff
@@ -1,48 +1,95 @@
...
+      ref
+          .read(loginViewModelProvider.notifier)
+          .checkEmail(emailController.text);
+    });
+    passwordController = TextEditingController();
+    passwordController.addListener(() {
+      ref
+          .read(loginViewModelProvider.notifier)
+          .checkPassword(passwordController.text);
+    });
```
Comment (manh-t) [inline lib/screens/login/login_form.dart:49]: we don't need the listeners too

File: lib/screens/login/login_view_model.dart [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    _email = email ?? '';
+    _isValidEmail
+        .add(!(email == null || email.isEmpty || !email.contains('@')));
+  }
+
+  void checkPassword(String? password) {
+    _password = password ?? '';
+    _isValidPassword
+        .add(!(password == null || password.isEmpty || password.length < 8));
+  }
```
Comment (manh-t) [inline lib/screens/login/login_view_model.dart:33]: modify the methods to return the error message if the value is invalid.
```suggestion
  String? checkEmail(String? email, String errorMessage) {
    _email = email ?? '';
    if (email == null || email.isEmpty || !email.contains('@')) {
      return errorMessage;
    }
    return null;
  }

  String? checkPassword(String? password, String errorMessage) {
    _password = password ?? '';
    if (password == null || password.isEmpty || password.length < 8) {
      return errorMessage;
    }
    return null;
  }
```

File: lib/screens/login/login_view_model.dart [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  late LoginUseCase loginUseCase;
+
+  String _email = '';
+  String _password = '';
+
+  final _isValidEmail = StreamController<bool>();
+  Stream<bool> get isValidEmail => _isValidEmail.stream;
+
+  final _isValidPassword = StreamController<bool>();
+  Stream<bool> get isValidPassword => _isValidPassword.stream;
```
Comment (manh-t) [inline lib/screens/login/login_view_model.dart:21]: then remove these variables

**Review by manh-t — CHANGES_REQUESTED** [review]
Currently the Login Form and its validation methods are complicated. I suggest following [the best practice](https://docs.flutter.dev/cookbook/forms/validation) from Flutter 🙋

File: pubspec.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaungkhantsoe) [reply pubspec.yaml:42]: Screenshot 2023-05-11 at 2 16 13 PM (2)

@manh-t I did fetch the pubspec again and also run `flutter analyze .`. It failed the check.

[the remark this answers is not in the document]
Comment (manh-t) [reply pubspec.yaml:42]: @kaungkhantsoe we can pass the Flutter lint easily fix it by adding a comment before:
```
// ignore: depend_on_referenced_packages
import 'package:collection/collection.dart';
```

**Review by manh-t — APPROVED** [review]
LGTM now 🚀

**Review by Thieurom — APPROVED** [review]
You're really fast, @kaungkhantsoe!
And thank you for helping taking over this task 🙏

**Pull request merged**